### PR TITLE
Refine parameterisation and add CAM interface

### PR DESCRIPTION
**Pull request by jatkinson1000**
This PR seeks to add an interface to the code parameterisation for the CAM model (part of CESM).

To do so we need to:
- [x] Interpolate relevant variables onto grid matching that required by parameterisation  
  We will interpolate based on pressure:
  SAM: pressure at middle of grid cell `pres` and interface `presi` from `sounding_z_pres0_tabs0.txt`
  - [x] This is incorrect - SAM cell-centres are in altitude space, not pressure space, so need to calculate p_mid!

  CAM: pressure at middle and interfaces of grid cells are `state%pmid` and `state%pint` 
  - [x] Normalise `p` by `p_surface` for both grids before interpolating ($\sigma$ coordinate space)
  - [x] Inputs: Use linear interpolation
  - [x] Outputs: Use first-order conservative regridding $^1$
- [x] Reformulate CAM variables into the form expected by the parameterisation
  - [x] Moisture:  
          SAM: `q` - total non-precipitating water **mixing ratio**
          CAM: `state%q(:,:,ixcldliq) + state%q(:,:,ixcldice) + state%q(:,:,1)` - cloud liquid + cloud ice + **specific humidity**
  - [x] Thermodynamics  
         SAM: `tabs_i` - absolute temperature
         CAM: `state%t(:,:)` - absolute temperature
  - [x] 'Constant' Profiles and Variables
          These need extracting from SAM to be used in CAM as 'constants', stored somewhere, and read in during initialisation routine
    - [x] `rho` from `rho from sounding_z_rho_rhow.txt`
    - [x] `z` from ~`sounding_z_rho_rhow.txt`~ `grd` (see discussion below)
    - [x] ~`pres` from `pres0` in `sounding_z_pres0_tabs0.txt` (in hPa)~ No longer required (see discussion below)
    - [x] `dz` in `setgrid.f90`
    - [x] `adz` in `setgrid.f90`
  - [ ] `y_in` needs setting somehow (will eventually be removed (?))
- [x] Reformulate the returned values to the form expected by CAM
  - [x] Moisture:
         SAM: Provides tendencies in `q`
         CAM: tendencies of specific humidity `ptend%q(:,:,1)`, cloud liquid water `ptend%q(:,:,ixcldliq)`, and cloud ice `ptend%q(:,:,ixcldice)` $^2$
  - [x] Thermodynamics:  
          SAM: Provides tendencies to `t` (normalised liquid ice static energy)
          CAM: `ptend%s` - tendency of dry static energy $^3$
  - [x] tendencies above parameterisation grid are `0.0`
- [x] Provide interfaces to the correct locations in the CAM code
        `physpkg.F90` rather than `convect_deep.F90`
- [ ] Handle other variables:
  - [ ] `cmfmc` - convective mass flux, to be set as `0.0` for now
  - [x] precipitation - work out what parts of precipitation are required  
          'Precipitation fluxes should be rescaled by the ratio of surface pressures for consistency'!?!
  - [ ] Detrained Liquid - unclear what to do here...

The above points will be expanded as we work through this PR.

May address #18 

1. Regridding by averaging over the overlapping parts of grid cells to improve conservation.

2. Lines 30-158 in `cloud.f90` convert `q` and `t` from SAM into temperature `tabs` and cloud liquid + cloud ice `qn`. Note that `cloud.f90` allows for `qp` being non-zero, but `qp` is zero in the parameterisation.  
Calculate water vapor as `qv = q - qn`.  
Convert `qn` into cloud liquid `qcc` and cloud ice `qii` using code in `statistics.f90`.  
Convert changes in `qv`, `qcc` and `qii` to tendencies `ptend%q(:,:,1)`, `ptend%q(:,:,ixcldliq)` and `ptend%q(:,:,ixcldice)`  by dividing by the timestep `dtn`.

3. Use `q` and `t` converted to tabs as above in [1.], then calculate change in temperature and convert to dry static energy tendency by multiplying by `cp` and dividing by the `dtn`.

### CODE REVIEW

**Comment by jatkinson1000** [comment]
Extracted SAM sounding file and written script to cobvert to netCDF.
Questions for meeting:
- Are `air_density` and `altitude` the correct CF variables?
- Do we want to flip the data so altitude in ascending order? Or not because that's how the Fortran wants it? What is CF convention on this?

**Comment by paogorman** [comment]
> * altitude in pressure and density soundings are different, is this correct or is one canonical? grd and pressure seem to match.

Yes, the altitude in the pressure sounding file is correct (the density sounding file has the same z but it is rounded).


> * Do we want to flip the data so altitude in ascending order? 

Yes, the data should be flipped.

> * Is [`SAMSON_NN/rf_production/grd`](https://github.com/Cambridge-ICCS/Neural_nework_parameterization/blob/main/SAMSON_NN/rf_production/grd) the correct file for the parameterisation to calculate dz and adz?

Yes, that is the right z file. I said in the meeting that you should just use the z in the pressure file, but it might be clearer to just use the 'grd' file and we then don't need the pressure file as discussed below.

With regard to the pressure file: this provides tabs0 and pres0 which are initial temperatures and pressure profiles. We don't use these in nn_convection_flux.f90. So I don't think we need the pressure file as long as we use the grd file to get z.

nn_convection_flux.f90 does use pres in a part of the code controlled by rf_uses_rh. But pres is the current pressure and not the initial pressure. Therefore, if we wanted to implement the rf_uses_rh option, it would have to happen in the SAM-CAM interface using the actual pressure from CAM. My advice is to just remove the rf_uses_rh option.

**Comment by jatkinson1000** [comment]
Thanks @paogorman, these should now be implemented in the latest commit (bumped script to v1.1).

**Comment by jatkinson1000** [comment]
Added routines to read sounding variables in from NetCDF to CAM interface.
Note: The sounding has more levels (`48`) than are required by the parameterisation (`nrf = 30`).
For now I read in the entire sounding as the parameterisation selects only the lowest `nrf` cells.
Users should be aware in the future to keep things consistent if developing a new NN from a new grid configuration.
I think this follows the element of least surprise as all that is required is to change `nrf` in the code and regenerate `SAM_sounding.nc` using the included tool from the appropriate simulation files.

**Comment by jatkinson1000** [comment]
Interpolation

- [ ] Interpolate relevant variables onto grid matching that required by parameterisation  
       We will interpolate based on pressure:
       SAM: pressure at middle of grid cell `pres` and interface `presi` from `sounding_z_pres0_tabs0.txt`
       CAM: pressure at middle and interfaces of grid cells are `state%pmid` and `state%pint` 
  - [ ] Normalise `p` by `p_surface` for both grids before interpolating ($\sigma$ coordinate space)
  - [ ] Inputs: Use linear interpolation
  - [ ] Outputs: Use first-order conservative regridding $^1$

@paogorman:
- I will look at generating a `presi` variable, but this would require the input files used in the runs that generated the NN. Could this be done running SAMSON with the following as a 'case': [SAMSON_NN/rf_production](https://github.com/Cambridge-ICCS/Neural_nework_parameterization/tree/main/SAMSON_NN/rf_production)?
- Do you know which variables are stored at cell centres and which at cell edges in SAM and CAM? I believe both are Arakawa-C, so presumably scalars (and hence everything we use) are at cell centres?
- As such we only need `presi` for the conservative regridding?
- For surface pressure is there a specific value that provides this, or do you mean the bottom element of the column pressure? And is the bottom cell centre sufficcient, or does it need to be the boundary value (i.e. `presi(1)` not `pres(1)`)?
- I'm still exploring the CAM code, but do you know the array structure of the CAM variables? Your document says e.g. `state%q(:,:,1)`, but should this be `state%q(:,:,k,1)` - i.e. a `lat, lon, alt/pres` grid/indexing system?

**Comment by paogorman** [comment]
> * I will look at generating a `presi` variable, but this would require the input files used in the runs that generated the NN. Could this be done running SAMSON with the following as a 'case': [SAMSON_NN/rf_production](https://github.com/Cambridge-ICCS/Neural_nework_parameterization/tree/main/SAMSON_NN/rf_production)?

I'm not sure which run this was, but any of the runs we did in this line of work should work as long as it is a cold start (because we want to get the initial profiles). I've attached a sounding from one of the cold-start runs I did:

[sounding_z_pres0_presi0_tabs0.txt](https://github.com/m2lines/convection-parameterization-in-CAM/files/12380760/sounding_z_pres0_presi0_tabs0.txt)

You could confirm it is consistent with the previous sounding file and the relation of presi and pres in pressz.f90.


> * Do you know which variables are stored at cell centres and which at cell edges in SAM and CAM? I believe both are Arakawa-C, so presumably scalars (and hence everything we use) are at cell centres?

For SAM, yes static energy (t), moisture and tendencies are at the mid levels (i.e. cell centres).  The fluxes are at the interfaces, but you are not interpolating those in the current approach.

For CAM I think you would need to double check this yourself. I don't think it should be tied to Arakawa C in particular since CAM has different dynamical cores. Relevant code is in dp_coupling which has different versions for different dynamical cores, but perhaps there is documentation somewhere about the vertical coordinate.


> * For surface pressure is there a specific value that provides this, or do you mean the bottom element of the column pressure? And is the bottom cell centre sufficcient, or does it need to be the boundary value (i.e. `presi(1)` not `pres(1)`)?

For SAM I would use the bottom element of presi. 

For CAM, I think you could use the bottom element of pint or the ps element of phys_state which presumably should be the same. It would be good to check they are the same.

> * I'm still exploring the CAM code, but do you know the array structure of the CAM variables? Your document says e.g. `state%q(:,:,1)`, but should this be `state%q(:,:,k,1)` - i.e. a `lat, lon, alt/pres` grid/indexing system?

The arrays in physpkg  have already been reshaped to only have a column index instead of separate lat/lon indices (this was done in dp_coupling.f90 which comes in different forms for the different CAM dynamical cores). That was the reason we needed our interface to accept both 2d or 3d arrays so it worked for both cam and sam.

> * You say in the above comment that we no longer need `pres0`, but do we not require it to do the interpolation, unless I misunderstand? If so I'll write it back in.

Yes, you're right! I'd forgotten why we needed pres0.

**Comment by paogorman** [comment]
A couple of other resources that may be helpful:

1) An older guide to coupling a column physics parameterization to CAM:
[cam_physics_interface_guide.pdf](https://github.com/m2lines/convection-parameterization-in-CAM/files/12381745/cam_physics_interface_guide.pdf)

2) Parts of what we are trying to do here in terms of coupling SAM to CAM was previously done in the spcam part of cesm:
my_cesm_sandbox/components/cam/src/physics/spcam/crm_physics.F90
my_cesm_sandbox/components/cam/src/physics/spcam/crm/crmx_crm_module.F90
It is good to be aware of this, but spcam is implemented in a sufficiently different way to what is needed for the NN scheme that I'm not sure how useful it will be in the end.

**Comment by jatkinson1000** [comment]
Questions:
- It's easy to flatten arrays (reshape), but not to add an axis (allocate). Therefore adjust the NN to work on 2D data and refactor SAM interface to flatten inputs?
- Can we do away with `tabs_i` and `q_i` as inputs and make them the same as `t` and `q` for CAM?
- Do we need to zero precipitation in the routine (based on `nstatis` in SAM)? Or is CAM handling this for us in `physbc`?
- should `dtn` be set to the CAM physics timestep (for now, do we anticipate subcycling the routine later at all?)?

**Comment by paogorman** [comment]
-It's easy to flatten arrays (reshape), but not to add an axis (allocate). Therefore adjust the NN to work on 2D data and refactor SAM interface to flatten inputs?

I think that is ok if necessary. (Could it be avoided using a pointer?) Using 2D arrays will probably slow down the NN parameterization when running in SAM, but perhaps that's not a large penalty.

-Can we do away with tabs_i and q_i as inputs and make them the same as t and q for CAM?

We should keep them so that we can run the same NN parameterization in SAM, but they can be set equal to t and q in the CAM specific interface to the neural net parameterization.

-Do we need to zero precipitation in the routine (based on nstatis in SAM)? Or is CAM handling this for us in physbc?

This bit of code is not needed for CAM and would not make sense (there is no nstatis in CAM). But the code would be needed for running the parameterization in SAM. So I suggest we remove it and later add a lightweight interface for running in SAM that takes care of this.

-should dtn be set to the CAM physics timestep (for now, do we anticipate subcycling the routine later at all?)?

Again it is fine to set dtn to dt for CAM, but dtn should be kept for when running in SAM to avoid confusion. Yes, there is a good chance we will need subcycling in CAM.

**Comment by jatkinson1000** [comment]
Thanks @paogorman , these were notes to myself for tomorrow but already answered!
As such I'll qualify - I wasn't suggesting we remove anything from the core routines, but was asking in terms of the CAM interface.
So setting `tabs_i` and `q_i` as a copy of `tabs` and `q` is good, `nstatis` I will make a dummy in terms of the CAM interfacing, and `dtn` is set to the CAM physics timestep, with the possibility of subcycling in the interface being added at a later point.
A follow up question/discussion for tomorrow would be about the treatment of precipitation in CAM.

**Comment by jatkinson1000** [comment]
For precipitation follow what ZM is doing.
Create a variable `yog_precip` (and `yog_snow = 0.0`)
We may need to interface this with a land model at a later date, but worry about this down the line.

**Comment by paogorman** [comment]
To follow cam convention, I think these would be precip_yog, snow_yog

**Comment by jatkinson1000** [comment]
Calculation of `tabs`, `qn`, and `qv` from SAM `t` and `q` implemented (testing required).
Questions around use of `if(domicroscaling)` and assumption that `qp == 0.0`.

Next steps are to:
- convert `qv` into `qcc` and `qii`.
- implement these conversions into the interface routine.
- Convert changes to tendencies (and convert SAM parameterisation to return changes, not tendencies).

CAM interfacing is ongoing:
- Note: Finish conversion of 2D array operations (and note to change SAM code)

**Comment by dorchard** [comment]
We will bring the `cp` (specific heat of dry air) value from SAM  (it does exist in some other places in CAM but not readily available to us at this point).

**Comment by jatkinson1000** [comment]
We want to run the ZM scheme to generate ZM tendencies as outputs, even when using YOG as the deep convection scheme.

Need to be careful about doing this in a way that allows calling `convect_deep_tend` before/after `convect_deep_tend_yog` but without 'double operating' on any variables. May require creating copies, or, more likely, returning copies, or, if lucky, nothing.

- [ ] Set up to call ZM and YOG routines and create tendencies for output in a safe way.

**Comment by paogorman** [comment]
> * [ ]  `t`, `tabs`, and `q_i` out of CAM

Should this be 'tabs', 'q_v', 'q_c', 'q_i'?
(where tabs means absolute temperature which is  state%t(:,:) in CAM.
See below for how (SAM) t is calculated.

> @paogorman order of interpolation/variable conversion steps could be swapped, but this seems to me to be the most sensible order that is likely to minimise numerical error, do you agree?

With the proposed order, you are first interpolating SAM variables from cam to sam grid, and then interpolating CAM variables from sam to cam grid. I would have thought it would be better to do the interpolation on the sam variables both times (so switch the order).
 
> @paogorman To update `t` we need to formulate some form of `t` at the start of the timestep from CAM data. I think we discussed this, but can't see it in my notes above. Did we say using `cp*tabs` or `s` do you recall?

You could calculate t (meaning liquid-ice static energy /cp) based on line 479 of diagnose.f90 in the SAMson code:

    omn  = max(0.,min(1.,(tabs(i,j,k)-tbgmin)*a_bg))
     omp  = max(0.,min(1.,(tabs(i,j,k)-tprmin)*a_pr))
    tabs(i,j,k) = t(i,j,k)-gamaz(k)+ &
              (fac_cond+(1.-omn)*fac_fus)*qn(i,j,k)

Note this is written as an expression for tabs but you can just move things around to get an expression for t.
Note I set qp to zero because we are not using qp.
Note that qn is sum of cloud water + cloud ice which is available from CAM.

**Comment by jatkinson1000** [comment]
Thanks @paogorman 

Yes, in my mind `_i` was a subscript where `i` is one of `i, c, v` - I should have clarified this given 'i is one of i'.

My thoughts were that it would be better to convert variables on their 'native' grid before interpolating.
Interpolating first could introduce noise that would then be amplified by a conversion?
Happy to do either - it probably doesn't matter too much, and the conversion/interpolation procedures are generic subroutines, so it should be easy enough to swap them.

Thanks for the reference to diagnose.f90 - that's what I was missing!

**Comment by paogorman** [comment]
Yes, that was a good question about how to calculate t. It is not something that is done in SAMson except for the initialization, but there is no cloud at that point. Fortunately I think you can use the code from diagnose.f90.

Regarding the interpolation/conversion order:  I would first try the interpolation on the SAM variables both times for consistency. Being consistent is important because (to the extent possible) we want the case with no change in SAM variables from the ML routine to give no change in CAM variables. That makes me wonder if our interpolation will be reversible in this sense - probably not.

**Comment by jatkinson1000** [comment]
Regarding variable interpolation, whilst writing tests I realise that, particularly for SAM -> CAM conversion it makes sense to operate on the SAM gridded data as discussed above.
This is because we make use of certain values defined on the SAM grid (e.g. `gamaz(k)`) which would need to be interpolated should we use a different grid.

**Comment by jatkinson1000** [comment]
Meeting 14/02/2024

Testing variable conversion discussed with @paogorman:
- Test a profile for specific humidity ensuring you stay away from saturation
  - estimate saturation using q_sat = 0.622 * (p_v_sat / (p – p_v_sat))
  - estimate p_v_sat using empirical f(T)
  - so in reality set q profile, calculate T, and check q isn't above q_sat
- Take some CAM columns from over the tropics and put through the code.
  - This will check interpolation as well as variable conversion

Q - Is bottom interface pressure equal to surface pressure?

@judithberner asked Julio who replied
> PSL is 'sea-level pressure'. It is equal to PS over ocean, but is extrapolated over land.  The PINT variable is the pressure at layer interfaces.  So, the bottom level of PINT is equal to PS (should be).  Just FYI there are two more pressure variables - PMID and DELP - the mid-layer pressure and pressure change across the layers respectively. PINT,PMID and DELP were introduced when the vertical coordinate was switched to a 'dry mass' coordinate - i.e. water mass not included.

So it would seem pint(1) is the sensible choice for "surface pressure".
Can always check the CAM output to be sure.

Making progress with FTorch with Will and CISL - they will install a binary of libtorch compatible with cuda 11.8 or 12.1 on the stack. Apparently is no spack build for libtorch.

**Comment by jatkinson1000** [comment]
@paogorman In the notes above I wrote that we normalise pressure to create a $\sigma$ coordinate by dividing by the pressures of each grid by their respective surface values.

However, I was wondering, would it make sense to use the same surface pressure value to normalise both grids, as otherwise we are misaligning the grids in 'unnormalised'/absolute space?

I can see an argument the other way, saying that SAM is only a reference and that normalised surface pressure should have $\sigma = 1.0$ in both models, and there is a danger that we end up with a value of $\sigma < 1.0$ at the surface.\
I suppose ideally one might train the NN model with $\sigma$ rather than p?

**Comment by paogorman** [comment]
We want to use the respective surface pressures to capture the influence of
the lower boundary. But note this NN was only trained over ocean. In the
longer run we will want to switch to Griffin’s NN which was trained on
sigma coords and over land (it just needs a bit more testing in sam which
Griffin is working on).

**Comment by jatkinson1000** [comment]
Comments from discussion on 1/3/24:

- Define tendencies from NN return variables and conversion back to CAM variables.
  - `t` and `q_i` are returned from variable conversion scheme, take difference from input values to interface from CAM and divide by `dtn`
- Decide tendencies from SAM to be assigned to:
  - physics/cam/physics_types.f90 derived data type `physics_ptend%q(:,:,i)` which takes $\frac{dq_i}{dt}$.
  - multiply `t` by `cp` and divide by timestep for `physics_ptend%s(:,:)`
- Decide how best to operate switches:
  - Add a `YOG` flag as currently implemented and a `YOGCOMP` for running both ZM and YOG?
  - Check YOG variables have been included in the output and PBuf (JWA thinks variable declarations are done, but need to assign data values once coupled).

**Comment by dorchard** [comment]
(sorry I don't know where my comment went on this before) 
we thought it might make sense to call the flag `YOGCOMP` at the moment from the beginning as it is running both `YOG` and `ZM` now (and later we can add the separate `YOG` flag for running it in isolation).

**Comment by jatkinson1000** [comment]
@paogorman Is a reasonable surface boundary condition for `qv`, `qc`, and `qi` `0.0`?
Alternatively I could set equal to the bottom cell value or extrapolate to surface?
This is required for interpolation to the SAM grid.

Another question - I have implemented the calculation of tendencies on the SAM grid before interpolation back to the CAM grid (see lines 189-213 [here](https://github.com/m2lines/convection-parameterization-in-CAM/blob/93b75f3e85fff01d8d2852b7019e40ed66e838cc/NN_module/nn_interface_CAM.f90#L189))
In theory there should be no difference, but as ever in numerical practice there will be.
My reasoning was that it is easier to implement interpolation at the top of the SAM grid for tendencies rather than variables, and that the SAM grid is likely to be finer meaning more detail will be retained doing the calculation here.

**Comment by dorchard** [comment]
(To be clear: linear extrapolation downward for `qv`)

**Comment by paogorman** [comment]
Yes, linear extrapolation down for all variables.

**Comment by jatkinson1000** [comment]
Yes, by '`q`' I mean all `q` - `qv`, `qc`, and `qi`, as discussed 😄

**Comment by dorchard** [comment]
Re the simplification that we applied to these lines:
https://github.com/m2lines/Neural_nework_parameterization/blob/f81f5f695297888f0bd1e0e61524590b4566bf03/sam_code_NN/nn_convection_flux.f90#L439-L442
removing the factors `dz * ... * 1/dz` 

There was however a suggestion we still need to have an additional factor of `1/dz` in application of the sedimenting flux at the surface. This was discussed before but there was a potential miscommunication of `dz` heard as `adz`. 

However, it should be noted that the calculating of surface precipitation in our updated code does include the `1/dz` factor
https://github.com/m2lines/convection-parameterization-in-CAM/blob/cam-profile-test/NN_module/nn_convection_flux.f90#L434-L436
@paogorman  - is this where you were thinking it should go? Are you advocating for a further `1/dz` factor? We may need to the check the dimensional analysis on this...

**Comment by paogorman** [comment]
There seems to be two problems with the precipitation:

1) Only the precipitation from cloud sedimentation is currently included
(and this is a very small part of the total). The code is missing the
equivalent of the following lines from Janni's nn_convection_flux routine
which add in the much bigger contribution from 'autoconversion':

do k=1, nrf
            precsfc(i,j) =
precsfc(i,j)-q_tend_tot(k)*adz(k)*dz*rho(k)*(1/dz)! removed the time step
mult because q_tend_tot is already mult
end do

2) Both of the updates to prec (from sedimentation and autoconversion) need
to be multiplied by dz relative to what Janni's code does. This is because
the precsfc variable is later multiplied by dz elsewhere in the SAM code
prior to outputting statistics. See these lines in statistics.f90 and note
the dz factor:

do j=1,ny
do i=1,nx
 precsfc(i,j)=precsfc(i,j)*dz/dt*3600./(nstatis+1.e-5)
end do
end do

**Comment by jatkinson1000** [comment]
@dorchard @paogorman

RE 1) On the standalone testing that we have been looking at over the last fortnight the precipitation includes both sedimentation and autoconversion, see:

https://github.com/m2lines/convection-parameterization-in-CAM/blob/5de9f0272cf070176ed812350f4eccae71178ab1/profile_test/nn_convection_flux.f90#L376-L383

My intention is to propagate the updates we have made there (and in CAM) back here once we are happy.

RE 2) That helps understand, thank you. I'll make this change now, and make sure it is commented in the source to explain.

**Comment by paogorman** [comment]
There may be a problem in the code you just linked to:

precsfc(i) = precsfc(i) - q_sed_flux(1) * irhoadzdz(1) * rho(k) * adz(k) !!
 *dtn/dz

rho(k) should be rho(1)?
adz(k) should adz(1)?

**Comment by jatkinson1000** [comment]
Ah, they are in my code!
Just realised this is something I had caught last week but hadn't pushed yet.

That, plus the factor of `dz` is now added here:

https://github.com/m2lines/convection-parameterization-in-CAM/blob/cam-profile-test/profile_test/nn_convection_flux.f90#L376-L384

**Comment by paogorman** [comment]
Looks good

**Comment by jatkinson1000** [comment]
I have just made commits https://github.com/m2lines/convection-parameterization-in-CAM/pull/44/commits/399612b1f5eb6cdbb5c20765907965f0ffdea33f and https://github.com/m2lines/convection-parameterization-in-CAM/pull/44/commits/0e8f6fc026c482f42c2b13bf194d31cfc750188e which bring the code in-line with the current implementation used within CAM.

For full details see the commit messages.

This also updates the testing routines.

Suggestion is that this is merged to be consistent with CAM (after a couple of further points below) so that Adeleke can work on writing tests for regridding and any further work can be done in more manageable chunks.
- [ ] Remove SourceMods approach and replace with the `yog_intr.f90` file and a link + documentation for implementation in CAM
- [ ] Ensure any required data files are present in the repository
- [ ] Add any required documentation here

File: README.md [hunk 1]
```diff
@@ -41,16 +43,22 @@ Long term developments of this project will seek to re-deploy more complex ML pa
...
+The directory contains the additional files or details to interface the code with the CAM atmospheric model
+as part of the CESM model suite. It also includes a link to an implementation in a fork of CAM.
+
 
 ## Contributing
 
-This repository is currently private as it is new and work in progress.
+Contributions to the repository are welcome, particularly from anyone seeking to implement the
+parameterisation in another atmospheric model.
+We welcome code and details for interfacing.
```
Comment (MarionBWeinzierl) [inline README.md:60]: What do you mean by "details for interfacing"? Interface specifications?

File: CAM_interface/yog_intr.F90 [hunk 2]
```diff
@@ -0,0 +1,232 @@
+module yog_intr
```
Comment (MarionBWeinzierl) [inline CAM_interface/yog_intr.F90:1]: not sure this is the best name, as it's not immediately clear to me what "intr" stands for, and therefore what this module is

File: NN_module/test_utils.f90 [hunk 3]
```diff
@@ -1,9 +1,12 @@
 module test_utils
 
-  use :: precision, only: sp, dp
+  ! use :: precision, only: sp, dp
```
Comment (MarionBWeinzierl) [inline NN_module/test_utils.f90:3]: can this line be removed? Not sure why the change to the two lines below is needed.

File: NN_module/test_utils.f90 [hunk 4]
```diff
@@ -1,9 +1,12 @@
...
 
   implicit none
 
+  integer, parameter :: sp = 4
```
Comment (MarionBWeinzierl) [inline NN_module/test_utils.f90:7]: If we not using the precision module, maybe a more telling name (not just two letters) would be better

File: NN_module/nn_interface_CAM.f90 [hunk 5]
```diff
@@ -292,7 +306,7 @@ subroutine interp_to_sam(p_cam, p_surf_cam, var_cam, var_sam, var_cam_surface)
             elseif (p_norm_cam(i, nz_cam) > p_norm_sam(nrf)) then
                 ! This should not happen as CAM grid extends to higher altitudes than SAM
                 ! TODO This check is run on every iteration - move outside
-                write(*,*) "CAM upper pressure level is lower than that of SAM: Stopping."
+!                 write(*,*) "CAM upper pressure level is lower than that of SAM: Stopping."
```
Comment (MarionBWeinzierl) [inline NN_module/nn_interface_CAM.f90:309]: Keeping this warning might make sense, as otherwise it's not clear why this is stopping? Or does it happen all the time?

Maybe address the todo above, and move it outside of the iteration

**Review by MarionBWeinzierl — CHANGES_REQUESTED** [review]
There are a couple of commented out bits, and changes where I don't understand why they are necessary. 

Overall, I think it is important to now merge this asap and move forward from there, as this PR is too big and has gone on for a long time.

**Comment by MarionBWeinzierl** [comment]
Should this PR be changed to "ready for review"?

File: CAM_interface/yog_intr.F90 (hunk 2 above)
Comment (jatkinson1000) [reply CAM_interface/yog_intr.F90:1]: This matches the naming system for CAM physics modules so I would be inclined to stay consistent.
e.g. clubb_intr.F90, carma_intr.F90
See - https://github.com/m2lines/CAM-ML/tree/CAM-ML/src/physics/cam

Comment (jatkinson1000) [reply CAM_interface/yog_intr.F90:1]: There is a docstring below to add context: https://github.com/m2lines/convection-parameterization-in-CAM/blob/a515164a5ce32d49ed247f7096e566730b33d5fb/CAM_interface/yog_intr.F90#L1-L10

File: NN_module/test_utils.f90 (hunk 4 above)
Comment (jatkinson1000) [reply NN_module/test_utils.f90:7]: So, this is worth a discussion.
I would prefer to use sp and dp from iso_env_fortran.
However, since CAM is an older code it uses `real(8)`.
So to compile we need to be consisyent - see commit mesage on https://github.com/m2lines/convection-parameterization-in-CAM/pull/44/commits/399612b1f5eb6cdbb5c20765907965f0ffdea33f

Maybe we can discuss this in person?

File: NN_module/test_utils.f90 (hunk 3 above)
Comment (jatkinson1000) [reply NN_module/test_utils.f90:3]: See comment below. https://github.com/m2lines/convection-parameterization-in-CAM/pull/44/files/0e8f6fc026c482f42c2b13bf194d31cfc750188e..a515164a5ce32d49ed247f7096e566730b33d5fb#discussion_r1718099425

File: NN_module/test_utils.f90 (hunk 4 above)
Comment (dorchard) [reply NN_module/test_utils.f90:7]: (I was just thinking that `sp` and `dp` are reasonable common Fortran idioms. At least some comment could be added here to explain).

Comment (jatkinson1000) [reply NN_module/test_utils.f90:7]: After some discussion over this with Kacper I think we perhaps use the iso_fortran_env, include the precision module, and make it the responsibility of the user to set it appropriately.

This will increase portability, and "adding lots of files" is perhaps a weak argument against sensibly architectured and abstracted software.

Comment (jatkinson1000) [reply NN_module/test_utils.f90]: https://github.com/m2lines/convection-parameterization-in-CAM/pull/44/commits/632aa145ba116c8e3053149a4fdfe3eb8f8c2d36 Should have implemented this. It will now be a requirement for those using the scheme to set precision either in the precision file, or set in each file as suits them.

File: NN_module/test_utils.f90 (hunk 3 above)
Comment (jatkinson1000) [reply NN_module/test_utils.f90]: Should be covered now, again see comment below.

File: README.md (hunk 1 above)
Comment (jatkinson1000) [reply README.md]: I have clarified in https://github.com/m2lines/convection-parameterization-in-CAM/pull/44/commits/033cb038fdc514b7c5d1c8f8a0577a52dc3f8f90

File: NN_module/nn_interface_CAM.f90 (hunk 5 above)
Comment (jatkinson1000) [reply NN_module/nn_interface_CAM.f90]: I have reimplemented the warning (not expected to ever occur in reality) but left the restructure as a todo in future in the interests of merging this sooner. https://github.com/m2lines/convection-parameterization-in-CAM/pull/44/commits/b2395819081112daf916e75abcfa9134b014a295

**Review by MarionBWeinzierl — APPROVED** [review]

**Review by MarionBWeinzierl — APPROVED** [review]

**Review by dorchard — COMMENTED** [review]
Some minor things but otherwise great.

File: NN_module/nn_convection_flux.f90 [hunk 6]
```diff
@@ -116,137 +58,121 @@ end subroutine nn_convection_flux_init
...
-            !! delta values of t and q to be returned by the scheme
+        real(dp), dimension(size(tabs_i, 1), size(tabs_i, 2)) :: t_delta_adv, q_delta_adv, &
+                                                             t_delta_auto, q_delta_auto, &
+                                                             t_delta_sed, q_delta_sed
+            !! delta values of t and q generated by the NN
 
         != unit kg :: t_rad_rest_tend
-        real, intent(out), dimension(:,:,:) :: t_rad_rest_tend
-            !! tendency of t to be returned by the scheme
+        real(dp), dimension(size(tabs_i, 1), size(tabs_i, 2)) :: t_rad_rest_tend
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_convection_flux.f90:139]: No, see subroutine definition.

File: NN_module/nn_interface_CAM.f90 [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+
+
+        nx = size(q, 1)
+        nz = size(q, 2)
+
+        do k = 1, nz
+          do i = 1, nx
+            q(i,k) = qv(i,k) + qc(i,k) + qi(i,k)
+
+            ! omp  = max(0.,min(1.,(tabs(i,k)-tprmin)*a_pr))
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/nn_interface_CAM.f90:588]: I would prefer to leave for now to be clear that we are not considering precipitation constituent from CAM.
I have added a comment to clarify this.

File: CAM_interface/yog_intr.F90 [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+   integer :: i
+   integer :: nstep
+   integer :: ixcldice, ixcldliq      ! constituent indices for cloud liquid and ice water.
+   integer :: lchnk                   ! chunk identifier
+   integer :: ncol                    ! number of atmospheric columns
+
+   real(r8) :: ftem(pcols,pver)       ! Temporary workspace for outfld variables
+
+   logical  :: lq(pcnst)
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply CAM_interface/yog_intr.F90:193]: Done in https://github.com/m2lines/convection-parameterization-in-CAM/pull/44/commits/109218a7a171ca58cfd587346bc945356172fe85

File: NN_module/test_cam_interface.f90 [hunk 9]
```diff
@@ -0,0 +1,444 @@
+module cam_tests
+  !! Module containing individual tests for the CAM ML code
+
+  !--------------------------------------------------------------------------
+  ! Libraries to use
+  use netcdf
+  use precision, only: dp
+
+  use nn_cf_net_mod, only: relu, nn_cf_net_init, nn_cf_net_finalize, net_forward
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/test_cam_interface.f90]: Removed in https://github.com/m2lines/convection-parameterization-in-CAM/pull/44/commits/109218a7a171ca58cfd587346bc945356172fe85

File: NN_module/test_cam_interface.f90 [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+  use nn_convection_flux_mod, only:   nn_convection_flux, nn_convection_flux_init, nn_convection_flux_finalize
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/test_cam_interface.f90]: Removed in https://github.com/m2lines/convection-parameterization-in-CAM/pull/44/commits/109218a7a171ca58cfd587346bc945356172fe85

File: NN_module/README.md [hunk 11]
```diff
@@ -15,10 +15,22 @@ Files:
...
-To run the tests (requires `ifort` or `gfortran` and `netcdf`):
-Edit `Makefile` to select the appropriate compiler (`ifort` or `gfortran`) and then build with:
-```bash
-make test
+To run the tests (requires `ifort` or `gfortran` compiler and and `netcdf` and
+`netcdf-fortran` libraries) edit `Makefile` to select the appropriate compiler
+(`ifort` or `gfortran`), set flags, and select files, and then build with:
+```
+make
 ```
```
[the remark this answers is not in the document]
Comment (jatkinson1000) [reply NN_module/README.md:36]: There is an option to build and run the test and the interface test suite, which have different commands. Since this is not the focus of this PR and we want to merge ASAP to allow progress on other issues I suggest moving this as something to be covered by #64 and #67 please.

File: NN_module/nn_interface_CAM.f90 (hunk 7 above)
[the remark this answers is not in the document]
Comment (dorchard) [reply NN_module/nn_interface_CAM.f90]: Thanks

**Review by dorchard — APPROVED** [review]
